internal/postgres/reconcile.go

```diff
@@ -258,3 +261,146 @@ func PodSecurityContext(cluster *v1beta1.PostgresCluster) *corev1.PodSecurityCon
+
+	// create the pg_upgrade Job
+	upgradeJob := &batchv1.Job{}
+	upgradeJob.ObjectMeta = naming.PGUpgradeJob(cluster)
+
+	// set labels and annotations
+	var labels, annotations map[string]string
+	labels = naming.Merge(cluster.Spec.Metadata.GetLabelsOrNil(),
+		cluster.Spec.Upgrade.Metadata.GetLabelsOrNil(),
+		naming.PGUpgradeJobLabels(cluster.Name))
+	annotations = naming.Merge(cluster.Spec.Metadata.GetAnnotationsOrNil(),
+		cluster.Spec.Upgrade.Metadata.GetAnnotationsOrNil(),
+	)
+	upgradeJob.ObjectMeta.Labels = labels
+	upgradeJob.ObjectMeta.Annotations = annotations
+
+	certVolumeMount := corev1.VolumeMount{
+		Name:      naming.CertVolume,
+		MountPath: naming.CertMountPath,
+		ReadOnly:  true,
+	}
+	certVolume := corev1.Volume{
+		Name: certVolumeMount.Name,
+		VolumeSource: corev1.VolumeSource{
+			Projected: &corev1.ProjectedVolumeSource{
+				// PostgreSQL expects client certificate keys to not be readable
+				// by any other user.
+				// - https://www.postgresql.org/docs/current/libpq-ssl.html
+				DefaultMode: initialize.Int32(0o600),
+				Sources: []corev1.VolumeProjection{
+					{Secret: inClusterCertificates},
+					{Secret: inClientCertificates},
+				},
+			},
+		},
+	}
+
+	dataVolumeMount := DataVolumeMount()
+	dataVolume := corev1.Volume{
+		Name: dataVolumeMount.Name,
+		VolumeSource: corev1.VolumeSource{
+			PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+				ClaimName: inDataVolume.Name,
+				ReadOnly:  false,
+			},
+		},
+	}
+	//while true; do sleep 30; done;
+	script := fmt.Sprintf(`OLDVERSION=%d
+NEWVERSION=%d
+echo -e "Performing PostgreSQL upgrade from version $OLDVERSION to $NEWVERSION for cluster %s.\n"
+cd /pgdata
+
+echo -e "Step 1: Making new pgdata directory...\n"
+mkdir /pgdata/pg$NEWVERSION
+
+echo -e "Step 2: Initializing new pgdata directory...\n"
+/usr/pgsql-$NEWVERSION/bin/initdb -k -D /pgdata/pg$NEWVERSION
+
+echo -e "\nStep 3: Setting the expected permissions on the old pgdata directory...\n"
+chmod 700 /pgdata/pg$OLDVERSION
+
+echo -e "Step 4: Copying shared_preload_libraries setting to new postgresql.conf file...\n"
+grep shared_preload_libraries /pgdata/pg$OLDVERSION/postgresql.conf >> /pgdata/pg$NEWVERSION/postgresql.conf
+
+echo -e "Step 5: Running pg_upgrade check...\n"
+time /usr/pgsql-$NEWVERSION/bin/pg_upgrade --old-bindir /usr/pgsql-$OLDVERSION/bin --new-bindir \
+/usr/pgsql-$NEWVERSION/bin --old-datadir /pgdata/pg$OLDVERSION --new-datadir /pgdata/pg$NEWVERSION --link --check
+
+echo -e "\nStep 6: Running pg_upgrade...\n"
+time /usr/pgsql-$NEWVERSION/bin/pg_upgrade --old-bindir /usr/pgsql-$OLDVERSION/bin --new-bindir \
+/usr/pgsql-$NEWVERSION/bin --old-datadir /pgdata/pg$OLDVERSION --new-datadir /pgdata/pg$NEWVERSION --link
+
+echo -e "\nStep 7: Copying patroni.dynamic.json...\n"
+cp /pgdata/pg$OLDVERSION/patroni.dynamic.json /pgdata/pg$NEWVERSION
+echo -e "\npg_upgrade Job Complete!"
+`, cluster.Spec.Upgrade.FromPostgresVersion, cluster.Spec.PostgresVersion, cluster.Name)
```

Comment:
:seedling: This looks like one of those places where some additional validation could be beneficial before we run the Job, i.e. to ensure the "from version" is less than the "to version".

Can we add a note here indicating additional validation via webhook could be beneficial?

:question: Also, I assume right now the `--check` will simply fail if versions don't make sense (e.g. if the "from version" is greater than the "to version")?

Reply:
I dug into this a bit more and found that the current code, while preventing the upgrade from running, would allow the cluster to effectively be re-deployed with the wrong version. To handle this, I added some better use of conditions and events in `internal/controller/postgrescluster/instance.go`, which is where I put a TODO regarding webhook validation.

As for the `--check`, that's correct, the upgrade Job would fail and the logs would show:
```
Checking cluster versions                                   
This utility cannot be used to downgrade to older major PostgreSQL versions.
Failure, exiting
```
However, the current code should never allow that to happen.

Reply:
Sounds good, thank you!